src/translator/quality_estimator.cpp

```diff
@@ -0,0 +1,214 @@
+
+  AlignedVector<int16_t> A_prepared(featureMatrix.size());
+  AlignedVector<int16_t> B_prepared(modelMatrix_.size());
+
+  const auto width = getIntgenWidth(numFeatures_);
+  const auto modelMatrixColumn = getIntgenBColumn(1);
+
+  intgemm::Int16::PrepareA(featureMatrix.begin(), A_prepared.begin(), quant_mult, numWords, width);
+  intgemm::Int16::PrepareB(modelMatrix_.begin(), B_prepared.begin(), quant_mult, width, modelMatrixColumn);
+
+  AlignedVector<float> modelScores(numWords * modelMatrixColumn);
+
+  intgemm::Int16::Multiply(
+      A_prepared.begin(), B_prepared.begin(), numWords, width, modelMatrixColumn,
+      intgemm::callbacks::UnquantizeAndWrite(1.0f / (quant_mult * quant_mult), modelScores.begin()));
+
+  std::vector<float> wordQualityScores(numWords);
+
+  for (int i = 0; i < numWords; ++i) {
+    wordQualityScores[i] = modelScores[i * 8] + *intercept_;
+  }
+
+  return wordQualityScores;
+}
+
+AlignedVector<float> QualityEstimator::buildLinearModel() const {
+  AlignedVector<float> modelMatrix(getIntgenWidth(numFeatures_) * getIntgenBColumn(1));
+
+  for (auto& elem : modelMatrix) {
+    elem = 0.0;
+  }
+
+  for (int i = 0; i < numFeatures_; ++i) {
+    modelMatrix[i * 8] = *(coefficients_ + i);
+  }
+
+  return modelMatrix;
+}
+
+AlignedVector<float> QualityEstimator::extractFeatures(const ModelFeatures& modelFeatures) const {
+  const intgemm::Index numWords = modelFeatures.wordMeanScores.size();
+
+  AlignedVector<float> featureMatrix(numWords * getIntgenWidth(numFeatures_));
+
+  for (auto& elem : featureMatrix) {
+    elem = 0.0;
+  }
+
+  const int stdStart = 0;
+  const int meanStart = 1 * numFeatures_;
+  for (int i = 0; i < numWords; ++i) {
+    int j = 0;
+
+    for (const auto value : {modelFeatures.wordMeanScores[i], modelFeatures.minWordScores[i],
+                             static_cast<float>(modelFeatures.numSubWords[i]), modelFeatures.overallMean}) {
+      const auto stdsTemp = *(stds_ + j) != 0.0 ? *(stds_ + j) : 1.0f;
+      featureMatrix[8 * i * numFeatures_ + j] = (value - *(means_ + j)) / stdsTemp;
+      ++j;
+    }
+  }
+
+  return featureMatrix;
+}
+
+float QualityEstimator::computeWordProbabilities(std::vector<float>& wordQualityScores) const {
+  float sentenceScore = 0.0;
+
+  for (int i = 0; i < wordQualityScores.size(); ++i) {
+    auto& elem = wordQualityScores[i];
+    elem = 1 / (1 + std::exp(-elem));
+    sentenceScore += elem;
+  }
+  // TODO:
+  // Should we include or remove the end of sentence (EOS) token?
+  return sentenceScore / wordQualityScores.size();
+}
+
+QualityEstimator::WordsQualityEstimate QualityEstimator::computeQualityScores(const std::vector<float>& logProbs,
+                                                                              const AnnotatedText& target,
+                                                                              const size_t sentenceIdx) const {
+  auto [wordByteRanges, modelFeatures] = mapBPEToWords(logProbs, target, sentenceIdx);
```

Comment:
I'm not familiar with this syntax sugar, but smells like copy each time because of the wrap with `pair<...>`, could you please double-check?

Reply:
Hi @jerinphilip this a structured binding declaration. it's almolst like doing this:

auto pair = mapBPEToWords(logProbs, target, sentenceIdx)
const auto& wordByteRanges = pair.first;
const auto& modelFeatures = pair.second;

and because of the copy elision, it's will be only one construction, right?. 
If you think it's better the code above i will change.
References:
https://en.cppreference.com/w/cpp/language/structured_binding
https://en.cppreference.com/w/cpp/language/copy_elision
https://stackoverflow.com/questions/12953127/what-are-copy-elision-and-return-value-optimization

Reply:
Looks [ok](https://godbolt.org/z/4zxPPqojr). RVO seems to be working.